broadcast/src/linked/mod.rs

```diff
@@ -0,0 +1,640 @@
+//!
+//! 1. **Initializing** storage, networking, and cryptographic settings.
+//! 2. **Processing** incoming messages by validating data and signatures.
+//! 3. **Broadcasting** signed messages to all nodes.
+//! 4. **Verifying** message integrity and order.
+//! 5. **Updating Epochs** based on consensus.
+//!
+//! ## Key Design Principles
+//!
+//! - **Actor-Based Model**:  
+//!   The `Signer` (see `actor.rs`) encapsulates all functions, handling network I/O,
+//!   verification, timeouts, and state updates concurrently.
+//!
+//! - **Event Sourcing & Journaling**:  
+//!   Each new message is logged to a persistent journal. This maintains chain integrity and
+//!   enables reliable recovery after crashes.
+//!
+//! - **Asynchronous Processing**:  
+//!   Using Rust’s async/await, the actor processes tasks simultaneously, including epoch
+//!   refreshes, message rebroadcasts, and verification requests.
+//!
+//! ## Operational Flow
+//!
+//! 1. **Message Chaining**:  
+//!    Each new message includes a threshold signature over the previous message, forming an
+//!    immutable, verifiable chain.
+//!
+//! 2. **Signature Aggregation**:  
+//!    Validators produce partial signatures that combine into a threshold signature,
+//!    ensuring quorum endorsement.
+//!
+//! 3. **Epoch Management**:  
+//!    Epochs define the active set of signers and sequencers. The actor refreshes its epoch and
+//!    enforces strict bounds on accepted messages.
+//!
+//! 4. **Reliable Recovery**:  
+//!    Storage logs all sequencer messages, allowing nodes to replay messages after shutdown.
+//!
+//! 5. **Concurrent Operations**:  
+//!    The actor listens for network events, dispatches verification tasks, and manages timeouts
+//!    for efficient operation.
+//!
+//! # Public Modules
+//!
+//! - **`signer`**: Contains the actor that validates messages, links them, and handles threshold
+//!   signatures.
+//! - **`prover`**: Generates and verifies "proofs" (i.e. threshold signatures over messages).
+//!
+//! # Internal Modules
+//!
+//! - **`namespace`**: Generates namespace IDs for different signature types.
+//! - **`serializer`**: Serializes and deserializes messages.
+//! - **`wire`**: Contains the protocol buffer-generated message definitions.
+//! - **`mocks`**: Provides mock implementations for testing.
+
+use commonware_cryptography::PublicKey;
+
+mod namespace;
+mod serializer;
+
+#[cfg(test)]
+pub mod mocks;
+
+mod wire {
+    include!(concat!(env!("OUT_DIR"), "/wire.rs"));
+}
+
+pub mod prover;
+pub mod signer;
+
+/// `Epoch` is used as the `Index` type for the `Coordinator` trait.
+/// Defines the current set of sequencers and signers.
+pub type Epoch = u64;
+
+/// `Context` is used as the `Context` type for the `Application` and `Collector` traits.
+/// Stores a sequencer’s public key and sequential height, defining its unique position in the broadcast chain.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct Context {
```

Comment:
In consensus, I include `view` in context: https://github.com/commonwarexyz/monorepo/blob/8aa745f4c0bf6ddd41a96925990e8d474436fa36/consensus/src/threshold_simplex/mod.rs#L178-L191

I'm wondering if it makes sense to also include `epoch` (I don't expect the application to find it as useful as `view` in consensus but might be worth it...idk)?

Reply:
Good thought, I don't see a use for it at the moment, but it could be added pretty easily...